qiskit/circuit/variabletable.py

```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+"""
+Look-up table for varaible parameters in QuantumCircuit to support fast
+"""
+from collections import MutableMapping
+
+
+class VariableTable(MutableMapping):
+    """Class for managing and setting circuit variables"""
+
+    def __init__(self, *args, **kwargs):
+        """
+        the structure of _table is,
+           {var_object: [(instruction_object, parameter_index), ...]}
+        """
+        self._table = dict(*args, **kwargs)
+
+    def __getitem__(self, key):
+        return self._table[key]
+
+    def __setitem__(self, key, value):
+        """set the value of variable
+
+        Args:
+            key (Object): the variable to set
+            value (Number or dict): numeric constant or dictionary of
+                (Symbol:value) pairs
+        """
```

Comment:
Worth commenting here that the two operations bounds to `=` are very different, depending on whether or not `key` exists in the table. Might be worth splitting these into two separate methods, one to register a new variable->gate mapping, and another to assign values to a circuit variable.

Reply:
sounds good